src/Magento/FunctionalTestingFramework/StaticCheck/TestDependencyCheck.php

```diff
@@ -0,0 +1,385 @@
+use Magento\FunctionalTestingFramework\Test\Handlers\TestObjectHandler;
+use Magento\FunctionalTestingFramework\Test\Objects\ActionObject;
+use Magento\FunctionalTestingFramework\Util\ModuleResolver;
+use Magento\FunctionalTestingFramework\Util\TestGenerator;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\Finder\Finder;
+
+/**
+ * Class TestDependencyCheck
+ * @package Magento\FunctionalTestingFramework\StaticCheck
+ * @SuppressWarnings(PHPMD)
+ */
+class TestDependencyCheck implements StaticCheckInterface
+{
+    const EXTENDS_REGEX_PATTERN = '/extends=["\']([^\'"]*)/';
+    const ACTIONGROUP_REGEX_PATTERN = '/ref=["\']([^\'"]*)/';
+    const ACTIONGROUP_ARGUMENT_REGEX_PATTERN = '/<argument[^\/>]*name="([^"\']*)/';
+
+    /**
+     * Array of FullModuleName => [dependencies]
+     * @var array
+     */
+    private $allDependencies;
+
+    /**
+     * Array of FullModuleName => [dependencies], including flattened dependency tree
+     * @var array
+     */
+    private $flattenedDependencies;
+
+    /**
+     * Array of FullModuleName => PathToModule
+     * @var array
+     */
+    private $moduleNameToPath;
+
+    /**
+     * Array of FullModuleName => ComposerModuleName
+     * @var array
+     */
+    private $moduleNameToComposerName;
+
+    /**
+     * Transactional Array to keep track of what dependencies have already been extracted.
+     * @var array
+     */
+    private $alreadyExtractedDependencies;
+
+    /**
+     * Checks test dependencies, determined by references in tests versus the dependencies listed in the Magento module
+     *
+     * @param InputInterface $input
+     * @return string
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     */
+    public function execute(InputInterface $input)
+    {
+        MftfApplicationConfig::create(
+            true,
+            MftfApplicationConfig::UNIT_TEST_PHASE,
+            false,
+            false
+        );
+
+        ModuleResolver::getInstance()->getModulesPath();
+        if (!class_exists('\Magento\Framework\Component\ComponentRegistrar')) {
+            return "TEST DEPENDENCY CHECK ABORTED: MFTF must be attached or pointing to Magento codebase.";
+        }
+        $registrar = new \Magento\Framework\Component\ComponentRegistrar();
+        $this->moduleNameToPath = $registrar->getPaths(\Magento\Framework\Component\ComponentRegistrar::MODULE);
+        $this->moduleNameToComposerName = $this->buildModuleNameToComposerName($this->moduleNameToPath);
+        $this->flattenedDependencies = $this->buildComposerDependencyList();
+
+        $allModules = ModuleResolver::getInstance()->getModulesPath();
+        $filePaths = [
+            DIRECTORY_SEPARATOR . 'Test' . DIRECTORY_SEPARATOR,
+            DIRECTORY_SEPARATOR . 'ActionGroup' . DIRECTORY_SEPARATOR,
+            DIRECTORY_SEPARATOR . 'Data' . DIRECTORY_SEPARATOR,
```

Comment:
Why not include Page and Section here?

Reply:
Didn't think it was relevant, mostly. Section can't reference anything other than static selector strings, and while Pages can reference Sections, that linkage is not really used anywhere in MFTF.
If we do build an actual use for Page->Section linkage I think we should add that link check, but otherwise I don't see it as value adding as far as the check goes

Reply:
I see